network/p2p/p2pnode/gossipSubAdapter.go

```diff
@@ -0,0 +1,83 @@
+package p2pnode
+
+import (
+	"context"
+	"fmt"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/network/p2p"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// GossipSubAdapter is a wrapper around the libp2p GossipSub implementation
+// that implements the PubSubAdapter interface for the Flow network.
+type GossipSubAdapter struct {
+	gossipSub *pubsub.PubSub
+	logger    zerolog.Logger
+}
+
+var _ p2p.PubSubAdapter = (*GossipSubAdapter)(nil)
+
+func NewGossipSubAdapter(ctx context.Context, logger zerolog.Logger, h host.Host, cfg p2p.PubSubAdapterConfig) (p2p.PubSubAdapter, error) {
+	gossipSubConfig, ok := cfg.(*GossipSubAdapterConfig)
```

Comment:
given this must be a `*GossipSubAdapterConfig`, why not use that as the parameter type and avoid this check?

Reply:
`NewGossipSubAdapter()` is called by `defaultGossipSubFactory()` which expects a `GossipSubFactoryFuc` return type and that func expects the base type of `p2p.PubSubAdapterConfig`. Changing this to the sub-type would break creation of `GossibSubFactory`

https://github.com/onflow/flow-go/pull/3585/files#diff-4edd977ff4567bb1e4b7042a1d973e12b11732f4bb622bf934235fd30d444c59R126-R130